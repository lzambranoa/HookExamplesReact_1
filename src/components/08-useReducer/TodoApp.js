import React, { useEffect, useReducer } from 'react';
import './styles.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';



const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

   

    useEffect( () => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos]);

    const handleDelete = ( todoId ) => {
        console.log(todoId);

        //crear la action
        const action = {
            type: 'delete',
            payload: todoId
        }

        //dispatch
        dispatch(action);
    }

    const handleToogle = (todoId) => {


        dispatch({
            type: 'toogle',
            payload: todoId
        })
    }

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }



    return (
        <div>
            <h1>TodoApp ({todos.length}) </h1>
            <hr />

            <div className="row" >

                <div className="col-7" >
                    <TodoList 
                        todos={todos}
                        handleDelete={handleDelete}
                        handleToogle={handleToogle}
                    />
                </div>

                <div className="col-5" >
                   <TodoAdd 
                        handleAddTodo={handleAddTodo}
                   />
                </div>

            </div>



        </div>
    )
}