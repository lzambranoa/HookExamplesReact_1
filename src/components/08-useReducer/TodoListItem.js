import React from 'react';

export const TodoListItem = ({todo, index, handleDelete, handleToogle}) => {
    return (
        <li
        key={todo.id}
        className="list-group-item"
    > <p 
        onClick={ () => handleToogle(todo.id) }
        className={`${todo.done && 'complete'}`} 
    >
        {index + 1}. {todo.desc}</p>
        <button
        onClick={() => handleDelete(todo.id)}
        className="btn btn-danger" >
            Borrar
</button>
    </li>
    )
}