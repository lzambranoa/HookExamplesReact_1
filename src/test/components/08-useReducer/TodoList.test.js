import { shallow } from 'enzyme';
import React from 'react';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en el TodList', () => {

    const handleDelete = jest.fn();
    const handleToogle = jest.fn();

    const wrapper = shallow(
        <TodoList
            todos={demoTodos}
            handleDelet={handleDelete}
            handleToogle={handleToogle}
        />
    )

    test('Debe de mostrarse corectamente', () => {

        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de tener dos <TodoListItem />', () => {

        expect(wrapper.find('TodoListItem').length ).toBe( demoTodos.length );

        
    });
    test('la propiedad handleToogle debe ser una funcion', () => {

        expect( wrapper.find('TodoListItem').at(0).prop('handleToogle') ).toEqual( expect.any(Function) );
    });
})