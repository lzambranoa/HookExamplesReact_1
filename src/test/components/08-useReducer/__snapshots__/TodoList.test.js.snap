// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el TodList Debe de mostrarse corectamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleToogle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleToogle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
